File: serial/python/serial_terminal.py
# Copyright (c) 2025 by T3 Foundation. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#     https://docs.t3gemstone.org/en/license
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import select
import sys
import termios
import tty
from typing import Optional

import serial


class SerialPort:
    SUPPORTED_BAUD_RATES = [9600, 19200, 38400, 57600, 115200, 230400, 460800, 921600]

    def __init__(self):
        self._serial_connection: Optional[serial.Serial] = None

    def __del__(self):
        self.close()

    def configure(self, device: str, baud_rate: int) -> int:
        if baud_rate not in self.SUPPORTED_BAUD_RATES:
            print(f"Unsupported baud rate: {baud_rate}")
            print(f"Supported rates: {', '.join(map(str, self.SUPPORTED_BAUD_RATES))}")
            return 1

        try:
            self._serial_connection = serial.Serial(
                port=device,
                baudrate=baud_rate,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE,
                timeout=0.5,
                xonxoff=False,
                rtscts=False,
                dsrdtr=False,
            )
            return 0
        except serial.SerialException as e:
            print(f"Error opening serial port: {e}")
            return 1

    def close(self) -> None:
        if self._serial_connection and self._serial_connection.is_open:
            self._serial_connection.close()
            self._serial_connection = None

    def is_open(self) -> bool:
        return self._serial_connection is not None and self._serial_connection.is_open

    def get_fd(self) -> int:
        if self._serial_connection:
            return self._serial_connection.fileno()
        return -1

    def read(self, size: int = 1) -> bytes:
        if self._serial_connection:
            return self._serial_connection.read(size)
        return b""

    def write(self, data: bytes) -> int:
        if self._serial_connection:
            return self._serial_connection.write(data)
        return 0


class Terminal:
    def __init__(self):
        self._original_termios = None
        self.m_is_configured = False

    def __del__(self):
        self.restore()

    def configure(self) -> int:
        try:
            self._original_termios = termios.tcgetattr(sys.stdin.fileno())
            tty.setraw(sys.stdin.fileno())
            self.m_is_configured = True
            return 0
        except (termios.error, OSError) as e:
            print(f"Error setting terminal to raw mode: {e}")
            return 1

    def restore(self) -> None:
        if self.m_is_configured and self._original_termios:
            try:
                termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, self._original_termios)
                self.m_is_configured = False
            except (termios.error, OSError):
                pass


class SerialTerminal:
    def __init__(self):
        self.m_serial_port = SerialPort()
        self.m_terminal = Terminal()
        self.m_is_running = False

    def __del__(self):
        self.m_serial_port.close()
        self.m_terminal.restore()
        print("\nShutting down...")

    def initialize(self, device: str, baud_rate: int) -> int:
        if self.m_serial_port.configure(device, baud_rate):
            return 1

        print("=" * 46)
        print(f"port is     : {device}")
        print(f"baudrate is : {baud_rate}")
        print()
        print("Serial terminal started. Press Ctrl+C to exit.")
        print("=" * 46)

        if self.m_terminal.configure():
            return 1

        return 0

    def run(self) -> None:
        if not self.m_serial_port.is_open():
            return

        self.m_is_running = True
        serial_fd = self.m_serial_port.get_fd()
        stdin_fd = sys.stdin.fileno()

        while self.m_is_running:
            try:
                # Wait for input from either keyboard or serial port
                ready, _, _ = select.select([stdin_fd, serial_fd], [], [])

                # Handle keyboard input
                if stdin_fd in ready:
                    try:
                        data = sys.stdin.buffer.read(1)
                        if data:
                            # Check for Ctrl+C (ASCII 3)
                            if b"\x03" in data:
                                self.m_is_running = False
                                return

                            # Send to serial port
                            self.m_serial_port.write(data)
                    except (OSError, IOError):
                        print("error")
                        break

                # Handle serial port input
                if serial_fd in ready:
                    try:
                        data = self.m_serial_port.read(1)
                        if data:
                            sys.stdout.buffer.write(data)
                            sys.stdout.buffer.flush()
                    except (OSError, IOError):
                        break

            except KeyboardInterrupt:
                self.m_is_running = False
                break
            except select.error:
                break

    def stop(self) -> None:
        self.m_is_running = False

File: serial/cpp/serial_terminal.cpp
// Copyright (c) 2025 by T3 Foundation. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//     https://docs.t3gemstone.org/en/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

#include "serial_terminal.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <errno.h>
#include <fcntl.h>
#include <iostream>
#include <sys/select.h>
#include <unistd.h>

SerialPort::SerialPort() {}

SerialPort::~SerialPort()
{
    if (m_serial_fd >= 0)
    {
        ::close(m_serial_fd);
        m_serial_fd = -1;
    }
}

int SerialPort::configure(const std::string& device, int baud_rate)
{
    struct termios tty;
    speed_t speed;

    speed = get_baud_rate(baud_rate);
    if (speed == B0)
    {
        return 1;
    }

    m_serial_fd = ::open(device.c_str(), O_RDWR | O_NOCTTY | O_SYNC);
    if (m_serial_fd < 0)
    {
        perror("Error opening serial port");
        return 1;
    }

    if (tcgetattr(m_serial_fd, &tty) != 0)
    {
        perror("Error getting serial port attributes");
        ::close(m_serial_fd);
        m_serial_fd = -1;
        return 1;
    }

    cfsetospeed(&tty, speed);
    cfsetispeed(&tty, speed);

    // Configure 8N1 (8 data bits, no parity, 1 stop bit)
    tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8; // 8-bit chars
    tty.c_iflag &= ~IGNBRK;                     // disable break processing
    tty.c_lflag = 0;                            // no signaling chars, no echo,
                                                // no canonical processing
    tty.c_oflag = 0;                            // no remapping, no delays
    tty.c_cc[VMIN] = 0;                         // read doesn't block
    tty.c_cc[VTIME] = 5;                        // 0.5 seconds read timeout

    tty.c_iflag &= ~(IXON | IXOFF | IXANY); // shut off xon/xoff ctrl
    tty.c_cflag |= (CLOCAL | CREAD);        // ignore modem controls,
                                            // enable reading
    tty.c_cflag &= ~(PARENB | PARODD);      // shut off parity
    tty.c_cflag &= ~CSTOPB;                 // clear stop field
    tty.c_cflag &= ~CRTSCTS;                // no hardware flow control

    if (tcsetattr(m_serial_fd, TCSANOW, &tty) != 0)
    {
        perror("Error setting serial port attributes");
        ::close(m_serial_fd);
        m_serial_fd = -1;
        return 1;
    }

    return 0;
}

bool SerialPort::is_open() const
{
    return m_serial_fd >= 0;
}

int SerialPort::get_fd() const
{
    return m_serial_fd;
}

speed_t SerialPort::get_baud_rate(int baud) const
{
    switch (baud)
    {
    case 9600:
        return B9600;
    case 19200:
        return B19200;
    case 38400:
        return B38400;
    case 57600:
        return B57600;
    case 115200:
        return B115200;
    case 230400:
        return B230400;
    case 460800:
        return B460800;
    case 921600:
        return B921600;
    default:
        std::cout << "Unsupported baud rate: " << baud << std::endl;
        std::cout << "Supported rates: 9600, 19200, 38400, 57600, 115200, 230400, 460800, 921600" << std::endl;
        return B0;
    }
}

Terminal::Terminal() {}

Terminal::~Terminal()
{
    restore();
}

int Terminal::configure()
{
    struct termios raw;

    if (tcgetattr(STDIN_FILENO, &m_original_termios) == -1)
    {
        perror("Error getting terminal attributes");
        return 1;
    }
    m_is_configured = true;

    raw = m_original_termios;
    raw.c_lflag &= ~(ECHO | ICANON | IEXTEN | ISIG);
    raw.c_iflag &= ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON);
    raw.c_cflag |= CS8;
    raw.c_oflag &= ~OPOST;
    raw.c_cc[VMIN] = 0;
    raw.c_cc[VTIME] = 1;

    if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw) == -1)
    {
        perror("Error setting terminal to raw mode");
        return 1;
    }

    return 0;
}

void Terminal::restore()
{
    if (m_is_configured)
    {
        tcsetattr(STDIN_FILENO, TCSANOW, &m_original_termios);
        m_is_configured = false;
    }
}

SerialTerminal::SerialTerminal()
{
}

SerialTerminal::~SerialTerminal()
{
    m_terminal.restore();
}

int SerialTerminal::initialize(const std::string& device, int baud_rate)
{
    if (m_serial_port.configure(device, baud_rate))
    {
        return 1;
    }

    std::cout << "==============================================" << std::endl;
    std::cout << "port is     : " << device << std::endl;
    std::cout << "baudrate is : " << baud_rate << std::endl << std::endl;
    std::cout << "Serial terminal started. Press Ctrl+C to exit." << std::endl;
    std::cout << "==============================================" << std::endl;

    if (m_terminal.configure())
    {
        return 1;
    }

    return 0;
}

void SerialTerminal::run()
{
    fd_set read_fds;
    int max_fd;
    char buffer[256];
    ssize_t bytes_read;

    if (!m_serial_port.is_open())
    {
        return;
    }

    m_is_running = true;
    max_fd = (m_serial_port.get_fd() > STDIN_FILENO) ? m_serial_port.get_fd() : STDIN_FILENO;

    while (m_is_running)
    {
        FD_ZERO(&read_fds);
        FD_SET(STDIN_FILENO, &read_fds);
        FD_SET(m_serial_port.get_fd(), &read_fds);

        // Wait for input from either keyboard or serial port
        if (select(max_fd + 1, &read_fds, nullptr, nullptr, nullptr) < 0)
        {
            if (errno == EINTR)
            {
                continue;
            }
            break;
        }

        // Handle keyboard input
        if (FD_ISSET(STDIN_FILENO, &read_fds))
        {
            bytes_read = read(STDIN_FILENO, buffer, sizeof(buffer) - 1);
            if (bytes_read > 0)
            {
                // Check for Ctrl+C (ASCII 3)
                for (int i = 0; i < bytes_read; i++)
                {
                    if (buffer[i] == 3)
                    {
                        m_is_running = false;
                        return;
                    }
                }

                // Send to serial port
                if (write(m_serial_port.get_fd(), buffer, bytes_read) != bytes_read)
                {
                    break;
                }
            }
        }

        // Handle serial port input
        if (FD_ISSET(m_serial_port.get_fd(), &read_fds))
        {
            bytes_read = read(m_serial_port.get_fd(), buffer, sizeof(buffer) - 1);
            if (bytes_read > 0)
            {
                if (write(STDOUT_FILENO, buffer, bytes_read) != bytes_read)
                {
                    break;
                }
            }
            else if (bytes_read < 0 && errno != EAGAIN)
            {
                break;
            }
        }
    }
}

void SerialTerminal::stop()
{
    m_is_running = false;
}
